apache/trafficcontrol
# Prevent resource leak in API Tests WithObjs()

## What does this PR (Pull Request) do?


- [x] This PR fixes #5172 OR is not related to any Issue 

The issue was that the WithObjs functions to delete fixtures were deferred within a *for* loop, which is a resource leak. It's not an issue when running the API tests on more powerful machines, but the GitHub Actions virtual environment struggles with it.

In particular, `DeleteTestTypes()` calls `execSQL()`, so any test using `Types` (most of them) will hang in GitHub Actions if they fail.

This PR wraps a loop over `withFuncs[obj].Delete(t)` in a closure and defers that instead.

## Which Traffic Control components are affected by this PR?


- Traffic Ops
- CI tests

## What is the best way to verify this PR?

1. Observe the timeout occurring at 76196f143a (tests fail)
2. Observe the timeout not occurring at 1718b62076 (tests still fail)
3. See that the tests pass at a5214f8bd1 (reverts intentionally failing the API tests)

## If this is a bug fix, what versions of Traffic Control are affected?

- master (8434700b02)
- 4.1.0
- 3.1.0

## The following criteria are ALL met by this PR


- [x] This PR fixes tests
- [x] Only tests changed, no documentation necessary
